File: platform/polyglot/github_api/mjs/src/routes/security.mjs
/**
 * Security routes.
 * Exposes GitHub Security API operations as REST endpoints.
 * @module routes/security
 */

/**
 * Fastify plugin that registers security routes.
 * @param {import('fastify').FastifyInstance} fastify
 * @param {Object} opts
 * @param {import('../sdk/security/client.mjs').SecurityClient} opts.security - Security SDK client
 */
export default async function securityRoutes(fastify, opts) {
  const { security } = opts;

  /**
   * GET /repos/:owner/:repo/vulnerability-alerts - Get vulnerability alerts status.
   */
  fastify.get(
    '/repos/:owner/:repo/vulnerability-alerts',
    async (request, reply) => {
      const { owner, repo } = request.params;
      const result = await security.getVulnerabilityAlerts(owner, repo);
      return reply.send(result);
    },
  );

  /**
   * PUT /repos/:owner/:repo/vulnerability-alerts - Enable vulnerability alerts.
   */
  fastify.put(
    '/repos/:owner/:repo/vulnerability-alerts',
    async (request, reply) => {
      const { owner, repo } = request.params;
      await security.enableVulnerabilityAlerts(owner, repo);
      return reply.status(204).send();
    },
  );

  /**
   * DELETE /repos/:owner/:repo/vulnerability-alerts - Disable vulnerability alerts.
   */
  fastify.delete(
    '/repos/:owner/:repo/vulnerability-alerts',
    async (request, reply) => {
      const { owner, repo } = request.params;
      await security.disableVulnerabilityAlerts(owner, repo);
      return reply.status(204).send();
    },
  );

  /**
   * GET /repos/:owner/:repo/rulesets - List rulesets.
   */
  fastify.get('/repos/:owner/:repo/rulesets', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await security.listRulesets(owner, repo);
    return reply.send(result);
  });

  /**
   * GET /repos/:owner/:repo/rulesets/:id - Get a ruleset.
   */
  fastify.get(
    '/repos/:owner/:repo/rulesets/:id',
    async (request, reply) => {
      const { owner, repo, id } = request.params;
      const result = await security.getRuleset(owner, repo, parseInt(id, 10));
      return reply.send(result);
    },
  );

  /**
   * POST /repos/:owner/:repo/rulesets - Create a ruleset.
   */
  fastify.post('/repos/:owner/:repo/rulesets', async (request, reply) => {
    const { owner, repo } = request.params;
    const result = await security.createRuleset(owner, repo, request.body);
    return reply.status(201).send(result);
  });

  /**
   * PUT /repos/:owner/:repo/rulesets/:id - Update a ruleset.
   */
  fastify.put(
    '/repos/:owner/:repo/rulesets/:id',
    async (request, reply) => {
      const { owner, repo, id } = request.params;
      const result = await security.updateRuleset(
        owner,
        repo,
        parseInt(id, 10),
        request.body,
      );
      return reply.send(result);
    },
  );

  /**
   * DELETE /repos/:owner/:repo/rulesets/:id - Delete a ruleset.
   */
  fastify.delete(
    '/repos/:owner/:repo/rulesets/:id',
    async (request, reply) => {
      const { owner, repo, id } = request.params;
      await security.deleteRuleset(owner, repo, parseInt(id, 10));
      return reply.status(204).send();
    },
  );

  return Promise.resolve();
}

File: platform/polyglot/github_api/py/github_api/routes/security.py
"""
Security routes for the FastAPI server.

Maps HTTP endpoints to SecurityClient methods for vulnerability alerts
and repository rulesets.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Request

__all__ = ["router"]

router = APIRouter(tags=["security"])


def _get_security_client(request: Request):
    """Get the SecurityClient from app state."""
    from github_api.sdk.security import SecurityClient
    return SecurityClient(request.app.state.github_client)


@router.get("/repos/{owner}/{repo}/vulnerability-alerts")
async def get_vulnerability_alerts(
    request: Request, owner: str, repo: str
) -> dict[str, Any]:
    """Check if vulnerability alerts are enabled."""
    client = _get_security_client(request)
    return await client.get_vulnerability_alerts(owner, repo)


@router.put("/repos/{owner}/{repo}/vulnerability-alerts")
async def enable_vulnerability_alerts(
    request: Request, owner: str, repo: str
) -> dict[str, Any]:
    """Enable vulnerability alerts for a repository."""
    client = _get_security_client(request)
    return await client.enable_vulnerability_alerts(owner, repo)


@router.delete("/repos/{owner}/{repo}/vulnerability-alerts")
async def disable_vulnerability_alerts(
    request: Request, owner: str, repo: str
) -> dict[str, Any]:
    """Disable vulnerability alerts for a repository."""
    client = _get_security_client(request)
    return await client.disable_vulnerability_alerts(owner, repo)


@router.get("/repos/{owner}/{repo}/rulesets")
async def list_rulesets(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """List repository rulesets."""
    client = _get_security_client(request)
    params = dict(request.query_params)
    return await client.list_rulesets(owner, repo, params=params or None)


@router.get("/repos/{owner}/{repo}/rulesets/{ruleset_id}")
async def get_ruleset(
    request: Request, owner: str, repo: str, ruleset_id: int
) -> dict[str, Any]:
    """Get a specific repository ruleset."""
    client = _get_security_client(request)
    return await client.get_ruleset(owner, repo, ruleset_id)


@router.post("/repos/{owner}/{repo}/rulesets")
async def create_ruleset(request: Request, owner: str, repo: str) -> dict[str, Any]:
    """Create a repository ruleset."""
    client = _get_security_client(request)
    body = await request.json()
    return await client.create_ruleset(owner, repo, body)


@router.put("/repos/{owner}/{repo}/rulesets/{ruleset_id}")
async def update_ruleset(
    request: Request, owner: str, repo: str, ruleset_id: int
) -> dict[str, Any]:
    """Update a repository ruleset."""
    client = _get_security_client(request)
    body = await request.json()
    return await client.update_ruleset(owner, repo, ruleset_id, body)


@router.delete("/repos/{owner}/{repo}/rulesets/{ruleset_id}")
async def delete_ruleset(
    request: Request, owner: str, repo: str, ruleset_id: int
) -> dict[str, Any]:
    """Delete a repository ruleset."""
    client = _get_security_client(request)
    return await client.delete_ruleset(owner, repo, ruleset_id)
